File: src/main/java/com/shotvibe/shotvibelib/ConditionVar.java
package com.shotvibe.shotvibelib;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ConditionVar {
    public ConditionVar() {
        mLock = new ReentrantLock();
        mCondition = mLock.newCondition();
    }

    public void lock() {
        mLock.lock();
    }

    public void unlock() {
        mLock.unlock();
    }

    public void await() {
        try {
            mCondition.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public void signal() {
        mCondition.signal();
    }

    public void signalAll() {
        mCondition.signalAll();
    }

    private Lock mLock;
    private Condition mCondition;
}

File: src/main/java/com/shotvibe/shotvibelib/ConditionVar.m
#include "ConditionVar.h"

@implementation SLConditionVar
{
    NSCondition *condition_;
}


- (id)init
{
    if (self = [super init]) {
        condition_ = [[NSCondition alloc] init];
    }
    return self;
}


- (void)lock
{
    [condition_ lock];
}


- (void)unlock
{
    [condition_ unlock];
}


- (void)await
{
    [condition_ wait];
}


- (void)signal
{
    [condition_ signal];
}


- (void)signalAll
{
    [condition_ broadcast];
}


@end
